keps/sig-api-machinery/4568-resilient-watchcache-initialization/README.md

```diff
@@ -0,0 +1,462 @@
+-->
+
+<!--
+This question should be filled when targeting a release.
+For Alpha, describe what tests will be added to ensure proper quality of the enhancement.
+
+For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
+https://storage.googleapis.com/k8s-triage/index.html
+-->
+
+- <test>: <link to test coverage>
+
+##### e2e tests
+
+Given we're only modifying kube-apiserver, integration tests are sufficient.
+
+
+### Graduation Criteria
+
+#### Beta
+
+- Feature implemented behind a feature gate
+- Feature is covered with unit and integration tests
+
+#### GA
+
+TODO
+
+### Upgrade / Downgrade Strategy
+
+The feature is purely in-memory so update/downgrade doesn't require any
+specific considerations. 
+
+### Version Skew Strategy
+
+Feature touches only kube-apiserver and coordination between individual
+instances is not needed.
+
+## Production Readiness Review Questionnaire
+
+
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+- [X] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: ResilientWatchCacheInitilization
+  - Components depending on the feature gate: kube-apiserver
+- [ ] Other
+  - Describe the mechanism:
+  - Will enabling / disabling the feature require downtime of the control
+    plane?
+  - Will enabling / disabling the feature require downtime or reprovisioning
+    of a node?
+
+###### Does enabling the feature change any default behavior?
+
+Yes:
+- kube-apiserver has a new PostStartHook that may block its readiness for longer time
+- some GET/LIST/WATCH requests may now return 429 error when storage layer is not
+  initialized (or is re-initializing) [429 error was possible before too, this change
+  is only extending the number of situations when this may happen]
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes, via disabling the feature-gate in kube-apiserver.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+The feature is purely in-memory so it will just work as enabled for the first time.
+
+###### Are there any tests for feature enablement/disablement?
+
+The feature is purely in-memory so feature enablement/disablement will not provide
+additional value on top of feature tests themselves.
+
+
+### Rollout, Upgrade and Rollback Planning
+
+###### How can a rollout or rollback fail? Can it impact already running workloads?
+
+In case of bugs:
+- kube-apiserver may not initialize correct due to buggy implementation of the
+  introduced post-start hook
+- kube-apiserver may incorrectly reject some requests with 429
+
+###### What specific metrics should inform a rollback?
+
+- kubeapiserver /readyz behaving visibly different than before
+- `apiserver_request_total` (especially watch for requests finished with 429 code)
```

Comment:
Be more prescriptive here to make it extra obvious to a cluster admin what to watch for?  E.g. "If an apiserver continues to respond to /readyz requests with X significantly after startup should have completed ..."

Reply:
PTAL